server/src/main/java/org/elasticsearch/persistent/PersistentTasksClusterService.java

```diff
@@ -247,6 +247,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
+    /**
+     * This unassigns a task from any node, i.e. it is assigned to a {@code null} node with the provided reason.
+     *
+     * Since the assignment executor node is null, the {@link PersistentTasksClusterService} will attempt to reassign it to a valid
+     * node quickly.
+     *
+     * @param taskId           the id of a persistent task
+     * @param taskAllocationId the expected allocation id of the persistent task
+     * @param reason           the reason for unassigning the task from any node
+     * @param listener         the listener that will be called when task is unassigned
+     */
+    public void unassignPersistentTask(final String taskId,
+                                       final long taskAllocationId,
+                                       final String reason,
+                                       final ActionListener<PersistentTask<?>> listener) {
+        clusterService.submitStateUpdateTask("unassign persistent task from any node", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                PersistentTasksCustomMetaData.Builder tasksInProgress = builder(currentState);
+                if (tasksInProgress.hasTask(taskId, taskAllocationId)) {
+                    logger.trace("Unassigning task {} with allocation id {}", taskId, taskAllocationId);
+                    return update(currentState, tasksInProgress.reassignTask(taskId, new Assignment(null, reason)));
+                } else {
+                    if (tasksInProgress.hasTask(taskId)) {
```

Comment:
I'm wondering if the log traces are really useful, since the ResourceNotFoundException should appear in the log anyway?